core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceConcurrencyTest.java

```diff
@@ -0,0 +1,311 @@
+    ListeningScheduledExecutorService executor = newTestExecutor();
+    TestDataSupplier supplier = new TestDataSupplier(false);
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "a:b:c",
+            supplier,
+            AuthType.PASSWORD,
+            new TestCredentialFactory(),
+            executor,
+            keyPairFuture,
+            newRateLimiter());
+
+    assertThat(supplier.counter.get()).isEqualTo(0);
+    Thread.sleep(100);
+    SslData data = instance.getSslData();
+    assertThat(data).isNotNull();
+  }
+
+  @Test
+  public void testRefreshWhenRefreshRequestAlwaysFails() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = newTestExecutor();
+
+    InstanceDataSupplier supplier =
+        (CloudSqlInstanceName instanceName,
+            AccessTokenSupplier accessTokenSupplier,
+            AuthType authType,
+            ListeningScheduledExecutorService exec,
+            ListenableFuture<KeyPair> keyPair) -> {
+          ListenableFuture<?> f =
+              exec.submit(
+                  () -> {
+                    throw new RuntimeException("always fails");
+                  });
+          f.get(); // this will throw an ExecutionException
+          return null;
+        };
+
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "a:b:c",
+            supplier,
+            AuthType.PASSWORD,
+            new TestCredentialFactory(),
+            executor,
+            keyPairFuture,
+            newRateLimiter());
+
+    Thread.sleep(500);
+    Assert.assertThrows(RuntimeException.class, () -> instance.getSslData());
+    // Note: refresh attempts will continue the background. This instance.getSslData() throws an
+    // exception after the first refresh attempt fails.
+  }
+
+  @Test(timeout = 45000) // 45 seconds timeout in case of deadlock
+  public void testForceRefreshDoesNotCauseADeadlock() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = newTestExecutor();
+    TestDataSupplier supplier = new TestDataSupplier(true);
+    List<CloudSqlInstance> instances = new ArrayList<>();
+
+    final int instanceCount = 5;
+
+    for (int i = 0; i < instanceCount; i++) {
+      instances.add(
+          new CloudSqlInstance(
+              "a:b:instance" + i,
+              supplier,
+              AuthType.PASSWORD,
+              new TestCredentialFactory(),
+              executor,
+              keyPairFuture,
+              newRateLimiter()));
+    }
+
+    // Get SSL Data for each instance, forcing the first refresh to complete.
+    instances.forEach(i -> i.getSslData());
```

Comment:
This can be simplified to a lambda reference.

Reply:
Fixed.